### PR TITLE
openafs: Refactor and introduce server module

### PR DESCRIPTION
**Pull request by spacefrogg**
###### Motivation for this change
- Support newer kernels
- Add an openafs server systemd service to serve afs on a nixos server
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by spacefrogg** [comment]
Don't merge yet. The openafs package is completely unprepared to act as a server. Re-titled it as work in progress

**Comment by spacefrogg** [comment]
Consider this done. Please merge.

I added a new nixos module for the openafs server and changed the generation of config files to be copied at activation time. This is necessary, as the server must be able to write to its config location.

File: pkgs/tools/system/netdata/default.nix
Comment (Mic92) [inline pkgs/tools/system/netdata/default.nix]: I guess this commit was included by accident.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (Mic92) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: could be `NetInfo` also a symlink?

```
system.activationScripts.openafServerVar = mkIf (netInfo != "") (stringAfter [ "var" ] ''
  mkdir -p /var/openafs
  ln -sf "${netInfo}" /var/openafs/NetInfo
''
```

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (Mic92) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: Is the comment at the end a left-over? Would it be possible to create a symlink by assigning the path of the config files to `environment.etc."openafs/server".source` instead of copying them at start up time?

Comment (spacefrogg) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: The comment is an editor hint to not mark the rest of the file as a comment. (It recongnizes /* as the start of a comment). Moved it away.
For linking vs. copying see below.

**Comment by spacefrogg** [comment]
### Copying config files versus linking
The openafs server (the bosserver process, to be specific) checks access rights for some important configuration files and directories. It in particular wants write permissions to its own `/etc/openafs/server` directory and, of course, to /var/openafs. That's why, these directories cannot be served from the store.

Runtime changes to the server are also reflected through the configuration files (like adding a temporary server to the network, etc.), which is why the files have to be individually writable and cannot be served as links.

### Alternative approach to configuration
The only real configuration file (in the NixOS sense) is `/var/openafs/NetInfo`. The `/etc/openafs/server` directory is regarded as runtime data, and the server supports means to change these files remotely, as they have to be kept consistent between several servers in the same AFS cell. This also means, that it comes with its own tools to modify the files (namely the `bos` command). An alternative approach would be to completely ignore the configuration and leave it to the server. The NixOS approach to "purity" goes amiss here anyway, as the configuration is not dependent on this particular NixOS machine but on the position of the server inside the AFS cell.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (spacefrogg) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: It cannot. See below.

**Comment by spacefrogg** [comment]
@Mic92, @maggesi, @joachifm and @globin: Reviving this pull request.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (Mic92) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: I don't know much about openafs. How can I generate these files, when testing your service?

Comment (spacefrogg) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: As the names suggest these are key files, which is kerberos terminology for an equivalent of a password stored in a file. You can set up a kerberos server. For a kerberos realm EXAMPLE.COM you create a principal (equiv. to a login) named afs/example.com@EXAMPLE.COM in kerberos with a random password. `kadmin.local addprinc -randkey afs/example.com@EXAMPLE.COM`. And export it with `kadmin.local ktadd -k /etc/openafs/server/rxkad.keytab afs/example.com@EXAMPLE.COM`. I am not sure about the exact syntax but along these lines... The lowercase example.com is the AFS cell. After storing that the afs server should be able to start. (You need a set up and running kerberos server, still)

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: The construction of `netInfo` reads poorly for me, can it be restructured somewhat?
It seems it'll produce either an empty string or a derivation, which is irregular.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: Please add a type.

**Review by joachifm — CHANGES_REQUESTED** [review]

**Comment by joachifm** [comment]
Are there any user-visible changes that might need a changelog?

**Comment by spacefrogg** [comment]
I updated the pull request and addressed the remarks. The generation of the NetInfo file is optional and now produces a null-value, instead of the empty string, if unneeded.

I added the types accordingly. The problem with showing the default value of the bosConfig option is, that defaultText must not depend on derivations, which is not possible. I copied the text of the derivation and quoted the reference to the openafs package.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: This, too, needs `defaultText`; derivations in general render poorly and cause lots of unnecessary work just to build the manual. It's unfortunate that it can't be handled better, but that's how it is right now.

File: pkgs/servers/openafs/default.nix
Comment (joachifm) [inline pkgs/servers/openafs/default.nix]: If you care to, these raw calls to `compareVersions` have more readable wrappers in `versionAtLeast` & `versionOlder` and friends (see lib/strings.nix).

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: Seems like this could be simplified by using `writeText`

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: Does it matter if this runs before or after the "openafs" activation script? Could it not be a `preStart` to the server unit?

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: Here, `afsConfig` is copied to `/etc`, but in `preStart` the store path is referenced; if the latter is sufficient, why copy at all? Seems a little inconsistent.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (spacefrogg) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: It could be done in `preStart`. I wasn't able yet to figure out, what to do in `activationScripts` and when to use `preStart` of a service. Is `preStart` preferred for services?

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (spacefrogg) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: What are you referring to? I don't use `preStart` up to now. And yes, it has to be copied. The server checks prior start if it is able to write its own configuration files. The openafs server inherently expects to be the owner of its configuration as it supports online reconfiguration that is reflected in those files. To make matters worse, the openafs server owns files in `/etc/openafs` **and** `/etc/openafs/server` while the client wants to read its own files from `/etc/openafs`. So, both services must be able to copy/link files to those directories independently from each other, which is not easily achievable with the normal `environment.etc` approach. Neither the client nor the server can be redirected to use other configuration locations during runtime. (It could be set during build time but not independently for the server and the client.)

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: @spacefrogg one consideration that can help you decide is whether it makes sense to perform the action on *every* system activation. I'd say that `preStart` or a designated "init" service unit is strongly preferred.

File: nixos/modules/services/network-filesystems/openafs-server.nix
Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: I'm referring to https://github.com/spacefrogg/nixpkgs/blob/58fd7339bdacc728e9ce619fbd6ce70478f72ed3/nixos/modules/services/network-filesystems/openafs-client.nix#L83

Comment (spacefrogg) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: I could refer to the copy inside `/etc/openafs` in `preStart`, but I could not leave out the copying to `/etc/openafs`. The other client-side tools, like `aklog` etc., need access to the config files in the compile time-specified location, which is `/etc/openafs`.

It all boils down to the following two phenomena:
1. Both, the client and the server need access to a single compile time-specified location, where they reference disjunct sets of config files.
2. The server needs writing access to its own configuration.

This situation cannot (to my knowledge) be solved with `environment.etc` linking, as no two sources can link into the same location in `/etc/static`. And I need write access which can also only be solved by copying files around.

Comment (joachifm) [inline nixos/modules/services/network-filesystems/openafs-server.nix]: Seems like there's an opportunity to simplify by making the client only use the stuff in Store and move the server state to `/var/lib` (the goal would be to minimise state & make it easier to run read-only `/etc`). Anyway, I think that's something that can be done later, if at all.

**Comment by spacefrogg** [comment]
I moved the copying stuff into the services' `preStart` section.

@joachifm As you were asking for a Changelog (wherever that would go to): There are (theoretically) user-visible changes for the openafs client users. This change drops support for kerberos4, which is outdated for more than 15 years now and known to be insecure. It was never working together with a kerberos 5 installation in the first place, as the installed binaries collided. Dropping support for kerberos 4 also got rid of the collision. In the rare case that there are users of a openafs client on NixOS that use kerberos 4 for authentication, it would be time to change, now.

**Comment by joachifm** [comment]
Re changelog, if you deem it appropriate, you could drop a note in https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-1703.xml

**Review by joachifm — APPROVED** [review]

**Comment by joachifm** [comment]
Anyway, I have nothing further on this, so if it works & nobody complains I suppose we can integrate.

**Comment by spacefrogg** [comment]
@joachifm, @Mic92 Is anything blocking this from being merged?

**Comment by Mic92** [comment]
@spacefrogg sorry I never set it up completely with kerberos etc, maybe @joachifm did.

**Comment by Mic92** [comment]
who is @z77z btw (packet maintainer)? Does not seem to be on github?

**Comment by spacefrogg** [comment]
I don't know him or if he is still active. I would be willing to add myself to the list of maintainers of this package, although, due to the mention-bot, I will be notified of future changes to these files anyway. Is there an official policy of taking maintainership?

**Comment by joachifm** [comment]
@spacefrogg if you want to be a maintainer of this package, then you should add yourself. That's pretty much it for policy on that front, as far as I'm aware :)

**Comment by spacefrogg** [comment]
@joachifm Could we merge this into 17.03? Is anything stopping this from being merged?

**Comment by joachifm** [comment]
It'd be easier if it didn't do so much in a single changeset.  I also still don't like how it dumps stuff into /etc. Anyway, I think @globin can make a determination about integrating into the release.

**Comment by disassembler** [comment]
@spacefrogg this has merge conflicts. Can you rebase and fix?

**Comment by spacefrogg** [comment]
Ready to merge.

**Comment by Mic92** [comment]
I tested it with:

```nix
services.openafsServer = {
  enable = true;
  cellServDB = [{
    ip = "127.0.0.1"; dnsname = "localhost";
  }];
  cellName = "localhost";
};
services.kerberos_server.enable = true;
```

However I need still need a keytab:
```console
$ sudo systemctl status openafs-server.service
● openafs-server.service - OpenAFS server
   Loaded: loaded (/nix/store/ghlkq2hnkv8pcyxiam9sjl7mb9jclllx-unit-openafs-server.service/openafs-server.service; enabled; vendor preset: enabled)
   Active: inactive (dead)
Condition: start condition failed at Fri 2018-02-02 09:17:45 GMT; 4min 39s ago
           └─ ConditionPathExists=/etc/openafs/server/rxkad.keytab was not met
```

How can I get a keytab for my local kerberos service?
Ideally the `enable` option should also mention that. 
At the moment the user is left in the dark.

**Comment by spacefrogg** [comment]
Setting up an AFS server is much more complicated than just enabling a nixos module. You need a running Kerberos 5 server setup, to start with. You have to create a kerberos principal (i.e. account) with a special name ("afs/<your-chosen-cell-name>") and export that to the specific keytab ("/etc/openafs/server/rxkad.keytab").

This is really involved and way beyond the scope of the option description. It is documented in the Quick Start Guide, which, now, gets installed with the openafs package. But generally speaking, somebody who wants to setup an AFS server knows that this guide exists and is already following it while setting up the nixos service.

To make a long story short: What do you suggest on where to put the pointer to the documentation? I'd be happy to fix that.

**Comment by spacefrogg** [comment]
Added a better description to the enable option.

**Comment by spacefrogg** [comment]
Added a minor update to support kernels up to 4.15

**Pull request merged**
